cmd/tke-installer/app/installer/expansion.go

```diff
@@ -0,0 +1,143 @@
+/*
+ * Tencent is pleased to support the open source community by making TKEStack
+ * available.
+ *
+ * Copyright (C) 2012-2021 Tencent. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the “License”); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an “AS IS” BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package installer
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"tkestack.io/tke/cmd/tke-installer/app/installer/constants"
+)
+
+const expansionDirName = "expansions/"
+const defaultExpansionDir = constants.DataDir + expansionDirName
+
+// expansion layout
+const expansionLayoutHooksDir = "hooks/"
+
+func (t *TKE) initExpansion(ctx context.Context) error {
+
+	if t.Config.CustomExpansionDir == "" {
+		t.Config.CustomExpansionDir = defaultExpansionDir
+		t.backup()
```

Comment:
This backup func only run when t.Config.CustomExpansionDir == "". Is it right?

Reply:
yes, and once it is set up by default, this step will never come.

Reply:
I think backup will write and update tke.json. If user input his or her CustomExpansionDir, why not write and update tke.json via backup func?